File: app/src/main/java/com/wojdor/hearthcards/data/session/UserSession.java
package com.wojdor.hearthcards.data.session;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import com.wojdor.hearthcards.R;
import com.wojdor.hearthcards.application.util.JsonParser;
import com.wojdor.hearthcards.domain.VersionInfo;

public final class UserSession {

    private static final String VERSION_INFO_KEY = "VERSION_INFO_KEY";

    private static UserSession instance;

    private final String localeKey;

    private SharedPreferences sharedPreferences;
    private JsonParser jsonParser;

    private UserSession(Context context) {
        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
        jsonParser = new JsonParser();
        localeKey = context.getString(R.string.settings_preferences_locale_key);
    }

    public static UserSession getInstance(Context context) {
        if (instance == null) {
            instance = new UserSession(context);
        }
        return instance;
    }

    public VersionInfo getVersionInfo() {
        String versionInfoJson = sharedPreferences.getString(VERSION_INFO_KEY, null);
        return jsonParser.fromJson(VersionInfo.class, versionInfoJson);
    }

    public void setVersionInfo(VersionInfo versionInfo) {
        String versionInfoJson = jsonParser.toJson(versionInfo);
        sharedPreferences.edit().putString(VERSION_INFO_KEY, versionInfoJson).apply();
    }

    public String getLocale() {
        return sharedPreferences.getString(localeKey, null);
    }

    public void setLocale(String locale) {
        sharedPreferences.edit().putString(localeKey, locale).apply();
    }
}

File: app/src/main/java/com/wojdor/hearthstonecards/data/session/UserSession.kt
package com.wojdor.hearthstonecards.data.session

import android.content.Context
import android.preference.PreferenceManager
import com.wojdor.hearthstonecards.R
import com.wojdor.hearthstonecards.application.util.JsonParser
import com.wojdor.hearthstonecards.domain.VersionInfo

class UserSession private constructor(context: Context) {

    private val localeKey by lazy { context.getString(R.string.settings_preferences_locale_key) }
    private val sharedPreferences by lazy { PreferenceManager.getDefaultSharedPreferences(context) }
    private val jsonParser by lazy { JsonParser() }

    var versionInfo: VersionInfo?
        get() {
            val versionInfoJson = sharedPreferences.getString(VERSION_INFO_KEY, null) ?: return null
            return jsonParser.fromJson(VersionInfo::class.java, versionInfoJson)
        }
        set(versionInfo) {
            val versionInfoJson = jsonParser.toJson(versionInfo)
            sharedPreferences.edit().putString(VERSION_INFO_KEY, versionInfoJson).apply()
        }

    var locale: String?
        get() = sharedPreferences.getString(localeKey, null)
        set(locale) = sharedPreferences.edit().putString(localeKey, locale).apply()


    var wasLanguageChanged: Boolean
        get() = sharedPreferences.getBoolean(WAS_LANGUAGE_CHANGED_KEY, false)
        set(wasLanguageChanged) = sharedPreferences.edit().putBoolean(WAS_LANGUAGE_CHANGED_KEY, wasLanguageChanged).apply()

    companion object {
        private const val VERSION_INFO_KEY = "VERSION_INFO_KEY"
        private const val WAS_LANGUAGE_CHANGED_KEY = "WAS_LANGUAGE_CHANGED_KEY"

        private var instance: UserSession? = null

        fun getInstance(context: Context): UserSession {
            if (instance == null) {
                instance = UserSession(context)
            }
            return instance as UserSession
        }
    }
}
